runtime/vm/jfr.cpp

```diff
@@ -616,27 +653,45 @@ initializeJFR(J9JavaVM *vm)
 		goto fail;
 	}
 	if (omrthread_monitor_init_with_name(&vm->jfrState.isConstantEventsInitializedMutex, 0, "Is JFR constantEvents initialized mutex")) {
-		goto done;
+		goto fail;
 	}
 
 	if (!VM_JFRWriter::initializaJFRWriter(vm)) {
 		goto fail;
 	}
 
-	rc = JNI_OK;
+	if (lateInit) {
+		/* Go through existing threads. */
+		J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+		while (NULL != walkThread) {
+			/* only initialize a thread once */
+			if (NULL != walkThread->jfrBuffer.bufferStart) {
+				U_8 *buffer = (U_8*)j9mem_allocate_memory(J9JFR_THREAD_BUFFER_SIZE, OMRMEM_CATEGORY_VM);
+				if (NULL == buffer) {
+					goto fail;
+				} else {
+					walkThread->jfrBuffer.bufferStart = buffer;
+					walkThread->jfrBuffer.bufferCurrent = buffer;
+					walkThread->jfrBuffer.bufferSize = J9JFR_THREAD_BUFFER_SIZE;
+					walkThread->jfrBuffer.bufferRemaining = J9JFR_THREAD_BUFFER_SIZE;
+				}
+			}
+
+			walkThread = J9_LINKED_LIST_NEXT_DO(vm->mainThread, walkThread);
+		}
+
+		jfrStartSamplingThread(vm);
+	}
+
 done:
+	rc = JNI_OK;
 	return rc;
 fail:
 	tearDownJFR(vm);
 	goto done;
 }
 
-/**
- * Shut down JFR.
- *
- * @param vm[in] the J9JavaVM
- */
-static void
+void
 tearDownJFR(J9JavaVM *vm)
 {
 	PORT_ACCESS_FROM_JAVAVM(vm);
```

Comment:
you should walk all threads and flush the buffer, then flush gloable meomry out to file. After, you should free all local buffers.

Reply:
`VM.stopJFR()` now calls `jfrDump()` first to flush all the thread buffers and write out the global buffer before `tearDownJFR()`.